src/pagetable.c

```diff
@@ -26,24 +26,38 @@ struct pageentry {
     unsigned avail:3;
 
     unsigned addr:20;
-};
+};  // 32 bits = 4 bytes
 
 struct pagetable {
     struct pageentry entry[ENTRIES_PER_TABLE];
 };
 
 struct pagetable *pagetable_create() {
+    // Each pagetable is one page (4096 bytes) with 2^10 entries
+    // Each process has a 4 MB large pagetable
     return (struct pagetable *)memory_alloc_page(1);
 }
 
 void pagetable_init(struct pagetable *p) {
-    unsigned i, stop;
-    stop = total_memory * 1024 * 1024;
+    uint32_t lower_half_memory_top = (2048 - 1) * 1024 * 1024;
+    uint32_t stop = total_memory * 1024 * 1024;
+    if (stop > lower_half_memory_top) {
+        // Cap stop at the 2GB boundary, to avoid direct-mapping the higher
+        // user space memory
+        stop = lower_half_memory_top;
+    }
+
+    uint32_t i;
     for (i = 0; i < stop; i += PAGE_SIZE) {
         pagetable_map(p, i, i, PAGE_FLAG_KERNEL | PAGE_FLAG_READWRITE);
     }
-    stop = (unsigned)video_buffer + video_xres * video_yres * 3;
-    for (i = (unsigned)video_buffer; i <= stop; i += PAGE_SIZE) {
+
+    // SL: in VirtualBox, vram is separate from ram, and no matter how much
+    // physical memory it has, video_buffer is always 0xe0000000.
+    // TODO (SL): Ensure video buffer is mapped into superviser mode even
```

Comment:
Is there a bug in jira for this? If so we should include the bug number here, ex: NUN-25